pkg/forwarder/transaction_container.go

```diff
@@ -0,0 +1,188 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package forwarder
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+type transactionStorage interface {
+	Serialize([]Transaction) error
+	Deserialize() ([]Transaction, error)
+}
+
+type transactionPrioritySorter interface {
+	Sort([]Transaction)
+}
+
+// transactionContainer stores transactions in memory and flush them to disk when the memory
+// limit is exceeded.
+type transactionContainer struct {
+	transactions               []Transaction
+	currentMemSizeInBytes      int
+	maxMemSizeInBytes          int
+	flushToStorageRatio        float64
+	dropPrioritySorter         transactionPrioritySorter
+	optionalTransactionStorage transactionStorage
+	mutex                      sync.RWMutex
+}
+
+func tryNewTransactionContainer(
+	maxMemSizeInBytes int,
+	flushToStorageRatio float64,
+	optionalDomainFolderPath string,
+	storageMaxSize int64,
+	dropPrioritySorter transactionPrioritySorter) (*transactionContainer, error) {
+	if maxMemSizeInBytes <= 0 {
+		return nil, nil
+	}
+	var storage transactionStorage = nil
+	var err error
+
+	if optionalDomainFolderPath != "" && storageMaxSize > 0 {
+		serializer := NewTransactionsSerializer()
+		storage, err = newTransactionsFileStorage(serializer, optionalDomainFolderPath, storageMaxSize)
+		if err != nil {
+			return nil, fmt.Errorf("Error when creating the file storage: %v", err)
+		}
+	}
+
+	return newTransactionContainer(dropPrioritySorter, storage, maxMemSizeInBytes, flushToStorageRatio), nil
+}
+
+func newTransactionContainer(
+	dropPrioritySorter transactionPrioritySorter,
+	optionalTransactionStorage transactionStorage,
+	maxMemSizeInBytes int,
+	flushToStorageRatio float64) *transactionContainer {
+	return &transactionContainer{
+		maxMemSizeInBytes:          maxMemSizeInBytes,
+		flushToStorageRatio:        flushToStorageRatio,
+		dropPrioritySorter:         dropPrioritySorter,
+		optionalTransactionStorage: optionalTransactionStorage,
+	}
+}
+
+// Add adds a new transaction and flush transactions to disk if the memory limit is exceeded.
+// The amount of transactions flushed to disk is control by
+// `flushToStorageRatio` which is the ratio of the transactions to be flushed.
+// Consider the following payload sizes 10, 20, 30, 40, 15 with `maxMemSizeInBytes=100` and
+// `flushToStorageRatio=0.6`
+// When adding the last payload `15`, the buffer becomes full (10+20+30+40+15 > 100) and
+// 100*0.6=60 bytes must be flushed on disk.
+// The first 3 transactions are flushed to the disk as 10 + 20 + 30 >= 60
+// If disk serialization failed or is not enabled, remove old transactions such as
+// `currentMemSizeInBytes` <= `maxMemSizeInBytes`
+func (tc *transactionContainer) Add(t Transaction) (int, error) {
```

Comment:
Same here for private (also apply to every method of `transactionContainer`)

Reply:
Code updated